### ISSUE TITLE
[v4] Feat: New output configuration

### ISSUE THREAD

**Issue by nchanged** [fuse-box/fuse-box#1756]
Hey guys, here is something to talk though, many people were complaining that the output config is way too opinionated, here is what I suggest:

```ts
/*
 * Window exposure is a mechanism to expose certain path of the code to window
 *
 *  A call will be applied to the bottom of the bundle
 *  e.g
 *
 *  window[$exposedVariableName] = FuseBox.import($fuseBoxPath)
 *  Where $fuseBoxPath should look like $package/$modulePath
 *
 *  For example of an exposed config for project path would look like:
 *  {
 *     exposedVariableName : "foobar",
 *     fuseBoxPath : "default/foo.module.ts
 *  }
 */
export interface IOutputWindowExposure {
  exposedVariableName: string;
  fuseBoxPath: string;
}

export interface IOutputBundleConfigAdvanced {
  bundle: string;
  windowExposure?: IOutputWindowExposure;
}

/**
 *  target bundle name could be a string or an advanced configuration
 *  with target and exposed objects
 */
export type IOutputBundleConfig = string | IOutputBundleConfigAdvanced;

export interface IOutputConfig {
  /**
   * Root is essential when making custom configuration
   * That's how we determine relative paths for webIndex (index.html)
   */
  root: string;

  /**
   * Isolated API works only for one bundle output.
   * e,g app : "./target.build.js"
   * And would mean that FuseBox will make an inclosed environment where
   * the API will have access to the outside world (window)
   * This is usually the case when making WebWorker bundles / browser builds for libraries
   */
  isolatedApi?: boolean;

  /**
   * You can customise how development bundles look like
   * Hashes are not available for development
   * Possible variations:
   *
   * For one bundle:
   * app : "./app.js"
   * app : "./app.js"
   * windowExposure : { fuseBoxPath : "default/foo", exposedVariableName : foobar }
   *
   * For two bundles
   *
   * app : "./app.js"
   * vendor : "./vendor.js"
   *
   */
  devBundles?: {
    app?: IOutputBundleConfig;
    vendor?: IOutputBundleConfig;
  };
  /**
   * The same as development and will be copied here unless the field is specified
   * Hashes are supported
   *
   * {
   *   app : "./app.$hash.js",
   *   vendor : "./vendor.$name.js"
   * }
   *
   * In addition you can do custom mapping to create additional bundles
   * {
   *    mapping : [
   *       matching : "moment*",
   *       target : "./vendor.moment.$hash.hs"
   *    ]
   * }
   *
   * Or with mapping and exposure
   * {
   *    mapping : [
   *       matching : "moment*",
   *       target : {
   *          bundle : "./vendor.moment.$hash.js",
   *          windowExposure : {
   *            fuseBoxPath : "moment/index",
   *            exposedVariableName : "moment"
   *          }
   *       }
   *    ]
   * }
   */
  prodBundles?: {
    app?: IOutputBundleConfig;
    vendor?: IOutputBundleConfig;
    mapping?: Array<{
      matching: string;
      target: IOutputBundleConfig;
    }>;
    cssSplitting?: boolean;
  };
}
```

So here is an example for a very custom output:

```ts
{
  root: './dist',
  devBundles: {
    app: './app.js',
    vendor: './vendor.js',
  },
  prodBundles: {
    app: './app.$hash.js',
    vendor: './vendor.$hash.js',
    mapping: [
      { matching: 'moment*', target: 'vendor.$hash.moment.js' },
      {
        matching: 'react*',
        target: {
          bundle: 'vendor.$hash.react.js',
          // that would mean that it won't have access to FuseBox scope
          // and can be re-used across projects
          // the only thing you need to do is to "shim" and avoid bundling it again
          isolatedApi: true,
          windowExposure: { fuseBoxPath: 'react/index', exposedVariableName: 'React' },
        },
      },
    ],
  },
};
```

Let me know what you think.

So what you can do:

* Let go and don't bother: FuseBox will **preconfigure** everything. 
* Customise every single bundle
* No hashes will be appended if not requred
* **Customise** exposed variables per bundle
* Create custom bundles with desired contents and expose those

**Comment by SephReed** [comment]
This all looks very good and logical.  I'm definitely going to use this to remove hashes from my bundle naming and instead add either "prod" or "dev" (ie `vendor.prod.js`).

I think this is a very good idea.  I do have a couple UX suggestions though:

1. I think that by default the file names should not include hashes, not because I think that's better, but because I think it's the least surprising/confusing outcome.  Also, hashes are a bit of an advanced feature that most beginning users will only trip over.  

2. I think the `devBundles` and `prodBundles` configurations would be more localized if they were part `runDev` and `runProd` respectively.  eg:
```
fuse.runDev({
  bundles: {
    app: './app.js',
    vendor: './vendor.js',
  },
});
```

Overall though, this seems like it will do a phenomenal job of giving users control over what bundles are made and where they are put.

**Comment by jeremy-coleman** [comment]
there may be an easier way, which is basically what the google devs added to next.js in v9. The jist is to just bundle everything from src into a single file (excluding all vendor files), then copy deps from node_modules to their own file, which is similar to a vendor single file but just not a single file. 

so like
dist/
 main.js
 vendors/
  moment.js
  react.js
  react-dom.js
 
It actually loads faster this way, bc downloads are multi threaded at the os level and the browser can concurrently parse. Users could do whatever post-processing they wanted with some very simple scripts too. There is also the benefit (undocumented i think) that es modules's COMPILED code (aka post-jit) can be shared across multiple iframes on the same origin, which is awesome. I did a test with lit html es imports from unpkg and spammed like 15-20 iframes , all importing lit, the heap memory didnt hardly move.

**Comment by SephReed** [comment]
This is a nice feature.  Perhaps it could be integrated with something like the following:

```
prodBundles: {
    app: './app.$hash.js',
    vendor: './vendor.$hash.js',
    mapping: [
      { matching: 'moment*', target: 'vendor.$hash.moment.js' },
      {
        matching: (path) => /node_modules/.test(path),
        target: (path) => {
          const moduleName = path.match(/.+/node_modules/([^/\.]+?)(\/.+)/)?/[1]; // the name of the folder
          bundle: `vendor/${moduleName}.js`,
          isolatedApi: true,
          windowExposure: { fuseBoxPath: `${moduleName}/index`, exposedVariableName: moduleName },
        },
      },
    ],
  },
```

**Comment by paxperscientiam** [comment]
@nchanged Would `windowExposure` have a sister option like, say, `globalExposure` for node? Also, what about environments where neither `window` nor `global` are available?

For instance, I'm not able to develop for Google Apps Script with v4 of `FuseBox` as there's no way to expose a function. I've worked around this, however, by using a polyfill for `globalThis`.

**Comment by nchanged** [comment]
All great ideas, thanks, guys!

@SephReed the only reason why there are hashes in the first now is that when `output` is left unconfigured we preconfigure it to `dist` folder with `app` and `vendor` files. So when you have multiple instances preconfigured they will all write to the same folder with the same file names.

I do like the idea of a localised configuration.
```ts
runDev({
   bundle : {}
})
```

Looks legit and makes sense!

@paxperscientiam  Good catch, I totally forgot about nodejs. How about we rename it to just  `expose` and since FuseBox knows the target it will pick the right function to make it happen.


@jeremy-coleman that's also nice for the output, as to what @SephReed is saying that should be applied in a different configuration field (that we need to discuss too)
 But let me understand it first.

So, instead of bundling the actual lib, we'd substitute it with a reference to a `window` object by injecting the script tag and rewriting the import statement 

e.g

```
import React from "react"
```

Would become at build time;
```
const React = window["React"].default
```

and the config would look something like that:

```ts
externals : [ {
   name : "react"
   script : "node_modules/......./react.vendor.js"
} ]
```

**Comment by jeremy-coleman** [comment]
@nchanged yea exactly. You could also optionally swap out the iife/umd bundle for an es import and do the module/nomodule approach. You could also drop the vendor folder all together and add links to unpkg or some other cdn. 

for config, i'm not sure externals is the place to set this up. I would think externals would be for ignoring the externals completely. Maybe another config option like 'bundleStrategy': 'single file' | 'split files' | 'cdn'.

**Comment by nchanged** [comment]
@jeremy-coleman but you can't combine externals and output in one bundle (or can you ?) You prepare the bundles first then you use externals which would override the statements , right ?

**Comment by FKobus** [comment]
Good start and good discussion so far!

Love the proposal from [SephReed](https://github.com/fuse-box/fuse-box/issues/1756#issuecomment-573182841)

### 1. Split output bundle into smaller bundles
I would like to propose the following in consideration with HTTP2 widely available and used by devs, that means that smaller bundle sizes actually matter.

```js
export interface IOutputBundleConfigAdvanced {
  bundle: string;
  maxBundleSize?: number;
  windowExposure?: IOutputWindowExposure;
}
```
By adding maxBundleSize, we split the output bundle into multiple bundles 

for example
```
maxBundleSize: 330

app.1.js  (329kb)
app.2.js  (328kb)
app.3.js  (128kb)
```

### 2. Multiple entry files
Sometimes we create a application that has multiple entries. With the way the config is setup now, we only support one entry file in fusebox. To achieve such a goal, we can take two routes.

#### 1. Entries as array
We can convert the app entry point to an more generic entries point that takes a single entry or an array of entries.
```ts
export interface IOutputConfig {
  prodBundles?: {
    entries?: IOutputBundleConfig | Array<IOutputBundleConfig>;
    ...
  }
}
```
Example:
```
const prodBundles = {
  entries: [
    './index.js',
    {
      bundle: './custom_entry.$hash.js'
    }
  ],
  vendor: 'vendor.$hash.js',
  mapping: [
    { matching: 'moment*', target: 'vendor.$hash.moment.js' },
    ...
  ],
}
```

Output in this case without configuration would be something like:

```
index.js
custom_entry.js
vendor.js
```

**Side effect**
In this case it's maybe necessary to add a bundleName config option to keep full control of the output?
Off course we can use the filename as output name, but a lot of people use `index.js/ts` as their main entry point (i think) and i don't think index.js as a bundle name is suitable right?

```ts
export interface IOutputBundleConfigAdvanced {
  bundle: string;
  outputName: string;
  windowExposure?: IOutputWindowExposure;
}
```

#### 2. Entries as object
While reading my suggestion, I came up with maybe a better way:

```ts
export interface BundleConfig {
  entries: {
    [key: string]: IOutputBundleConfig;
  };
  cssSplitting: boolean;
  mapping: Array<{
    matching: string | RegExp;
    target: IOutputBundleConfig;
  }>;
  vendor: IOutputBundleConfig;
}
```
Example:
```
const prodBundles = {
  entries: {
    app: './index.js',
    custom: {
      bundle: './custom_entry.$hash.js'
    }
  },
  vendor: 'vendor.$hash.js',
  mapping: [
    { matching: 'moment*', target: 'vendor.$hash.moment.js' },
    ...
  ],
}
```

Output in this case without configuration would be:

```
app.js
custom.js
vendor.js
```

#### 3. Extra keys are extra bundles
If we don't want to overcomplicate the structure by adding a configurable output name, we can do maybe something like this:

```ts
export type IOutputBundleConfig = {
  bundle: string;
  isolatedApi?: boolean;
  windowExposure?: {
    exposedVariableName: string;
    fuseBoxPath: string;
  };
  maxBundleSize?: number;
} | string;

export type BundleConfig = {
  [key: string]: IOutputBundleConfig;
} & {
  app: IOutputBundleConfig;
  cssSplitting: boolean;
  mapping: Array<{
    matching: string | RegExp;
    target: IOutputBundleConfig;
  }>;
  vendor: IOutputBundleConfig;
}
```
Example:
```
const prodBundles: BundleConfig = {
  app: './app.$hash.js',
  app_two: {
    bundle: './entry_two.js',
  },
  vendor: './vendor.$hash.js',
  mapping: [
    { matching: 'moment*', target: 'vendor.$hash.moment.js' },
    ...
  ],
};
```

In this case the output will be something like this:
```
app.js
vendor.js
app_two.js
```

### 3. Matching based on string and RegExp
Instead of a string matching on the mapping, I would suggest to add a RegExp. So we can type more convenient matching.

```js
mapping?: Array<{
  matching: string | RegExp;
  target: IOutputBundleConfig;
}>;
```

### 4. Commons bundle
I would like the feature where fuse-box automatically generates a commons bundle. Let's say we have multiple entry points that uses the same modules. These shared modules should be outputted into a commons bundle. Also, if we're heavily relying on code splitting to generate splitted bundles, we might want to put shared modules between a dynamic module and the main app into a commons bundle.

Giving this example:
```
'index.ts': `
  import { counter } from './utils';

  async function load() {
    const dynamicFunc = await import('./a');
  }
  counter();
`,
'a.ts': `
  import './b';
  export default function a() {}
`,
'b.ts': `
  import { counter } from './utils';
  counter();
`,
'utils.ts': `
  export function counter() {};
`,
}
```

In this case we can achieve a goal of the following:

**app.js (main)**
  - index.ts  

**a.js (split)**
  - a.ts
  - b.ts  

**commons.js**
  - utils.ts
 
It could operate the same as vendor, because it's a reserved word.

Example:
```js
const prodBundles = {
  app: './app.$hash.js',
  commons: {
    enable: true,
    maxBundleSize: 300
  }
};
```
or 
```js
const prodBundles = {
  app: './app.$hash.js',
  commons: true
};
```

Which could mean that the `IOutputBundleConfig` would look like something like this:
```ts
export type IOutputBundleConfig = {
  bundle: string;
  enable: boolean;
  isolatedApi?: boolean;
  windowExposure?: {
    exposedVariableName: string;
    fuseBoxPath: string;
  };
  maxBundleSize?: number;
} | string | boolean;
```

To be honest, I didn't completely think this last suggestion through, so feedback is welcome!

**Comment by jeremy-coleman** [comment]
@nchanged im just thinking maybe externals should exclude the file all together? for ex: 

this would be the output with externals: [],
```
dist/
 main.js
 vendors/
  moment.js
  react.js
  react-dom.js
```

and this would be with externals: ["moment": Moment]. Meaning the user would need to manually add a script link to unpkg, or wherever they want to get it from.
```
dist/
 main.js
 vendors/
  react.js
  react-dom.js
```

or it may not be necessarily at all

think in the case with externals: [],
```
dist/
 main.js
 vendors/
  moment.js
  react.js
  react-dom.js
```

and head will have <link src=vendor/moment.js />
the user could always just delete the vendors folder and replace with <link src=unpkg.com/moment/>

**Comment by paxperscientiam** [comment]
>  Good catch, I totally forgot about nodejs. How about we rename it to just expose and since FuseBox knows the target it will pick the right function to make it happen.

That sounds good to me! For clarification though, this would mean that something like `expose: "someArbitraryFunction"` would ensure that `someArbitraryFunction` is highest scope?

A suggestion for a potentially better way to specify exposed functions:
```ts
expose: {
`default/path/to/index.ts$someFunction$someOtherFunction`,
`default/path/to/other.ts$someOtherOtherFunction`,
Window: {
    `default/path/to/some/other/files.ts$someOtherOtherOtherFunction`,
 }
}
```
Does that make sense?

**Comment by SephReed** [comment]
I really like a lot of the feature ideas I'm seeing here about different ways of code bundling.  But if I could include a sort of user comprehension / future proofing general suggestion it would be this:

For anything beyond a simple bundling configuration, I think there must be a way for us to create a function/context relationship for bundling which would make all of these features (and more) possible at the same time.

----

AVOID: So, for example, if I had a special use case, there could be a special field for it:
```
fuse.runProd({
  bundles: {
    sephsUseCase: true;
    someOtherUseCase: true;
  }
});
```
But this field becomes one more thing to document and understand, and the more fields there are, the more complex their relationships become.  _How do these two use cases interact_?

----

RECOMMEND: A context/function paradigm would make support easier (no complex interactions between fields), user comprehension better (for noobs, don't look behind the curtain), and ultimately give more control to power users.

```
// Hasty example

IBunde {
  // whatever needs to be known about a bundle
  name: string;
  path: string;
  content: string;
}

IBundlingContext {
  // whatever we need to know about bundling in general
  bundles: Array<IBundle>;
}

IBundlingModuleContext {
  // whatever we need to know about the module
  path: string;
  sizeKb: number;
  isNodeModule: string;
  content: string;
}

fuse.runProd({
  bundles: (ctx: IBundlingContext, module: IBundlingModuleContext) => {
    // ideas so far:
    // A. size cap bundles into chunks
    // B. export every node_module as its own file
    // C. split entry points
    // D. modify exposure (I could really use this one)
    // E. Whatever else
  }
});
```

-----

This is really more of a structure than function suggestion, but IMO there's a point at which configuration becomes so complex, it might as well just be normal programming.  I think anything outside of simply changing the name of a bundle would be far better suited to a function/context solution.

**Comment by nchanged** [comment]
@SephReed I totally agree with you. In fact, I had in mind a concept of a `ModuleRouter`
What it is, in a nutshell, is:

```ts
moduleRouter.route(IncomingModule)
```

Where we do our generic magic, but we're also exposing an API to the users to play with it. e.g

fuse.runProd({
   bundles : {
     route : (module) => {
         return SomeBundle // or "string" or whatnot
     }
   }
})

The idea behind is to allow users to create their own desirable outputs without limitation. I think plugins could also be a good option

```ts
pluginOpinionatedOptimisedOutput()
```

To summarise, I'd keep the configuration simple aiming to meet the standards with splitting by size e.t.c but also exposing a powerful yet understandable mechanism to create anything.

**Comment by kwhitley** [comment]
Love this discussion, but in particular @FKobus comments on maxBundleSize for automatic splitting of the vendors.  This seems to in effect solve some of the other asks around vendor-specific split-outs to speed up transfer, as you can't load the app before they're all done, but really all you care about is getting those individual files down to a set size.  I'm sure splitting a single large vendor module in half for example isn't trivial, but it would certainly benefit the end devs/users!

**Comment by jeremy-coleman** [comment]
some other small points, chrome downloads up to 6 files in parallel (although i think http is spec'd up to 50, im not sure on that though). Chrome also has beneficial treatment for files that are between 30kb and 50kb, so that is the optimal size to shoot for.

**Comment by nchanged** [comment]
rest assured, `maxBundleSize` is coming ;-)

**Comment by FKobus** [comment]
> Chrome also has beneficial treatment for files that are between 30kb and 50kb, so that is the optimal size to shoot for.

@jeremy-coleman do you have a source for this statement? I'm curious what we can do to meet these requirements exactly

**Comment by charsleysa** [comment]
Closed due to inactivity.
